compiler/src/iree/compiler/Codegen/Common/GPU/GPULowerCoalescedDMAToGlobalLoads.cpp

```diff
@@ -0,0 +1,362 @@
+  int64_t totalBits = destType.getNumElements() * elementSize;
+
+  // Then, compute number of bytes each thread should load.
+  assert(totalBits % subgroupSize == 0 &&
+         "Total bits must be divisible by subgroup size");
+  int64_t bitsPerThread = totalBits / subgroupSize;
+
+  assert(bitsPerThread % maxLoadInstructionBits == 0 &&
+         "Bits per thread must be divisible by max load instruction bits");
+  int64_t copiesPerThread = bitsPerThread / maxLoadInstructionBits;
+  return copiesPerThread;
+}
+
+// Lowers iree_gpu.coalesced_gather_dma operations within scf.forall loops
+// with warp mapping to amdgpu.gather_to_lds operations inside scf.for loops.
+//
+// In short, looks for patterns like this:
+//   scf.forall (...) in (32, 1) {
+//     %1 = iree_gpu.coalesced_gather_dma %indices, %source into %dest
+//   } {mapping = [#gpu.warp<linear_dim_1>, #gpu.warp<linear_dim_0>]}
+//
+// And replaces with:
+//   %subgroup_id = gpu.subgroup_id
+//   %lane_id = gpu.lane_id
+//   scf.for %iv = %c0 to %num_copies step %c1 {
+//     %thread_index = affine.delinearize_index(%subgroup_id, ...)
+//     %store_index = affine.delinearize_index(%lane_id, ...)
+//     amdgpu.gather_to_lds %source[%thread_index] into %dest[%store_index]
+//   }
+//   gpu.barrier
+struct LowerCoalescedGatherDMAPattern
+    : public OpRewritePattern<IREE::GPU::CoalescedGatherDMAOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LowerCoalescedGatherDMAPattern(MLIRContext *context,
+                                 ArrayRef<int64_t> workgroupSize,
+                                 int64_t subgroupSize,
+                                 int64_t maxLoadInstructionBits)
+      : OpRewritePattern<IREE::GPU::CoalescedGatherDMAOp>(context),
+        workgroupSize(workgroupSize), subgroupSize(subgroupSize),
+        maxLoadInstructionBits(maxLoadInstructionBits) {}
+
+  LogicalResult matchAndRewrite(IREE::GPU::CoalescedGatherDMAOp dmaOp,
+                                PatternRewriter &rewriter) const override {
+    Location loc = dmaOp.getLoc();
+
+    LDBG() << "=== matchAndRewrite called for: " << dmaOp;
+
+    // Check if this DMA op is eligible for transformation.
+    if (failed(isEligibleForGlobalDMA(dmaOp))) {
+      LDBG() << "Op is not eligible for transformation";
+      return failure();
+    }
+
+    scf::ForallOp forallOp = dmaOp->getParentOfType<scf::ForallOp>();
+
+    // Replace the entire scf.forall with amdgpu.gather_to_lds.
+    rewriter.setInsertionPoint(forallOp);
+
+    auto destMemRefType = dyn_cast<MemRefType>(dmaOp.getInit().getType());
+    int64_t numCopiesPerThread = getNumOfGlobalLoadsPerThread(
+        destMemRefType, subgroupSize, maxLoadInstructionBits);
+
+    Value indices = dmaOp.getIndices();
+    auto indicesType = cast<MemRefType>(indices.getType());
+
+    Value dest = dmaOp.getInit();
+    auto destType = cast<MemRefType>(dest.getType());
+
+    Value subgroupId = rewriter.create<gpu::SubgroupIdOp>(
```

Comment:
Why do we need a subgroup ID here? ISn't the copy _already_ distributed to subgroups?